packages/eslint-plugin/src/rules/consistent-indexed-object-style.ts

```diff
@@ -0,0 +1,118 @@
+import { createRule } from '../util';
+import {
+  AST_NODE_TYPES,
+  TSESTree,
+} from '@typescript-eslint/experimental-utils';
+
+export default createRule({
+  name: 'consistent-indexed-object-style',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'Enforce or disallow the use of the record type',
+      category: 'Stylistic Issues',
+      // too opinionated to be recommended
+      recommended: false,
+    },
+    messages: {
+      preferRecord: 'A record is preferred over an index signature',
+      preferIndexSignature: 'An index signature is preferred over a record.',
+    },
+    fixable: 'code',
+    schema: [
+      {
+        enum: ['always', 'never'],
+      },
+    ],
+  },
+  defaultOptions: ['always'],
+  create(context) {
+    const sourceCode = context.getSourceCode();
+
+    if (context.options[0] === 'never') {
+      return {
+        TSTypeReference(node): void {
+          const typeName = node.typeName as TSESTree.Identifier;
+          if (typeName.name !== 'Record') {
+            return;
+          }
+
+          const params = node.typeParameters?.params;
+          if (params?.length !== 2) {
+            return;
+          }
+
+          context.report({
+            node,
+            messageId: 'preferIndexSignature',
+            fix(fixer) {
+              const key = sourceCode.getText(params[0]);
+              const type = sourceCode.getText(params[1]);
+              return fixer.replaceTextRange(
+                node.range,
+                `{ [key: ${key}]: ${type} }`,
+              );
+            },
+          });
+        },
+      };
+    }
+
+    /**
+     * Convert an index signature node to record code as string.
+     */
+    function toRecord(node: TSESTree.TSIndexSignature): string {
+      const parameter = node.parameters[0] as TSESTree.Identifier;
```

Comment:
Prefer actual checks to casts.
```suggestion
      const parameter = node.parameters[0];
      if (parameter.type !=== AST_NODE_TYPES.Identifier) {
        return;
      }
```

Reply:
The return type is used in the fixer. Your suggestion would potentially make the autofixer return undefined, which actually causes syntax errors. The check would need to happen before the node is even reported. I tried to look for such a situation, but I couldn’t find any.

Reply:
I would consider adjusting the type of the function to be `string | null` then, and specifically handling the `null` case in the fixer.

You can return `null` from a fixer and eslint will ignore it then

Reply:
I did this now, but it did cause code coverage to drop as I couldn’t find a situation that causes the index signature parameters to have a length other than 0, be of anohter type than an identifier, or not have a type annotation.

Reply:
It's okay if the code coverage isn't complete.
I use it as a guide for how well you've tested your code. The targets are set high to encourage contributors to do thorough testing.

I review the coverage report myself if it's red - I won't block a PR unless I think you've missed something.